rs/state_manager/tests/common/mod.rs

```diff
@@ -107,28 +109,37 @@ impl Verifier for RejectingVerifier {
 /// `size_limit`, applies `f` to the encoded slice and tries to decode the
 /// slice returned by `f`.
 ///
-/// If `destination_subnet` is `Some(subnet)` it will address the stream at
-/// `subnet` and at `SubnetId::new(42)` otherwise. The flag
-/// `should_pass_verification` allows to control whether the verification of the
-/// certification contained in the encoded `CertifiedStreamSlice` should succeed
-/// or not.
+/// The slice is encoded by `subnet_test_id(13)` and decoded by
+/// `subnet_test_id(42)`. If `destination_subnet` is `None`, the stream is
+/// addressed to the latter; if it is `Some(subnet)`, it is instead addressed to
+/// `subnet`. The `should_pass_verification` flag controls whether the
+/// verification of the certification of the encoded `CertifiedStreamSlice`
+/// should succeed or not.
 ///
 /// # Panics
 /// Whenever encoding and/or decoding would under normal circumstances return an
 /// error.
 pub fn encode_decode_stream_test<
-    F: FnOnce(StateManagerImpl, CertifiedStreamSlice) -> (StateManagerImpl, CertifiedStreamSlice),
+    F: FnOnce(&StateManagerImpl, CertifiedStreamSlice) -> CertifiedStreamSlice,
 >(
     stream: Stream,
     size_limit: Option<usize>,
     destination_subnet: Option<SubnetId>,
     should_pass_verification: bool,
     f: F,
 ) {
-    state_manager_test_with_verifier_result(should_pass_verification, |_metrics, state_manager| {
+    let sender_subnet = subnet_test_id(13);
+    let receiver_subnet = subnet_test_id(42);
+    state_manager_test_with_verifier_result(sender_subnet, true, |state_manager, _metrics, log| {
         let (_height, mut state) = state_manager.take_tip();
 
-        let destination_subnet = destination_subnet.unwrap_or_else(|| subnet_test_id(42));
+        if let Some(destination_subnet) = destination_subnet {
+            assert!(
+                destination_subnet != sender_subnet && destination_subnet != receiver_subnet,
+                "destination subnet, when specified, must be different from both sender and receiver subnets",
```

Comment:
Why do we require the destination subnet to be different from the receiver?

Reply:
As discussed on a different thread, either the tests using this want a `sender_subnet -> receiver_subnet` stream, actually encoded by `sender_subnet` and decoded by `receiver_subnet`, which is then valid on not based on how we tweak or not the slice.

The `destination_subnet` override is there to instead cause `sender_subnet` to encode the stream as being addressed to a different `destination_subnet` instead and then still decoded by `receiver_subnet`. This assert just ensures that the three subnets are actually different.